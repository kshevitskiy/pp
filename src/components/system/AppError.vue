<script lang="ts" setup>
import { apiError, ApiError } from '~/service'
import { computed, ref, watch } from 'vue'
import { AppStatus } from '~/types/enums'
import { delay } from '~/assets/utils'
import Alert from '~/components/common/Alert.vue'

type ErrorDataType<T> = {
  data: T
}

const DEFAULT_DURATION = 3000
const errors = ref<ErrorDataType<ApiError>[]>([])
const hasErrors = computed(() => errors.value.length > 0)

watch(
  () => apiError.value,
  (value) => {
    if (value) {
      errors.value.unshift({
        data: value
      })

      apiError.value = null
      hideErrors()
    }
  },
  {
    immediate: true
  }
)

async function hideErrors() {
  await delay(DEFAULT_DURATION)
  errors.value.pop()
}
</script>

<template>
  <Teleport to="#alert">
    <template v-if="hasErrors">
      <Alert
        v-for="(error, index) in errors"
        :key="`${index}-${error.data.status}`"
        :variant="AppStatus.DANGER"
        :collapsible="false"
      >
        <span class="font-medium">Unexpected error:</span>
        {{ error.data.message }}
      </Alert>
    </template>
  </Teleport>
</template>
