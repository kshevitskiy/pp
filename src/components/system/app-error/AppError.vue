<script lang="ts" setup>
import { ApiError } from '~/service'
import { computed, ref, watch } from 'vue'
import { AppStatus } from '~/types/enums'
import { delay } from '~/assets/utils'
import Alert from '~/components/common/alert/Alert.vue'

type ErrorDataType<T> = {
  data: T
}

const props = defineProps<{
  error: ApiError | null
}>()

const emit = defineEmits<{ (e: 'update:error', value: null): void }>()

const DEFAULT_DURATION = 3000
const errors = ref<ErrorDataType<ApiError>[]>([])
const hasErrors = computed(() => errors.value.length > 0)

defineExpose({ errors, hasErrors, hideErrors })

watch(
  () => props.error,
  (value) => {
    if (value) {
      errors.value.unshift({
        data: value
      })

      emit('update:error', null)
      hideErrors()
    }
  },
  {
    immediate: true
  }
)

async function hideErrors() {
  await delay(DEFAULT_DURATION)
  errors.value.pop()
}
</script>

<template>
  <Teleport to="#alert">
    <template v-if="hasErrors">
      <Alert
        v-for="(error, index) in errors"
        :key="`${index}-${error.data.status}`"
        :variant="AppStatus.DANGER"
        :collapsible="false"
      >
        <span class="font-medium">Unexpected error:</span>
        {{ error.data.message }}
      </Alert>
    </template>
  </Teleport>
</template>
