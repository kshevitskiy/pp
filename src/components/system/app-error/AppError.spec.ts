import { test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import AppError from './AppError.vue'

const errorFixture = {
  code: 400,
  message: 'Error message'
}

beforeEach(() => {
  const alert = document.createElement('div')
  alert.id = 'alert'
  document.body.appendChild(alert)
})

afterEach(() => {
  document.body.outerHTML = ''
})

test('mount component', async () => {
  const wrapper = mount(AppError, {
    props: {
      error: errorFixture
    }
  })
  expect(wrapper.html()).toMatchSnapshot()
})

test('find alert component', async () => {
  const wrapper = mount(AppError, {
    props: {
      error: errorFixture
    }
  })
  wrapper.findComponent({ name: 'Alert' })
})

test('updates errors ref', async () => {
  const wrapper = mount(AppError, {
    props: {
      error: errorFixture
    }
  })

  expect(wrapper.vm.errors).toHaveLength(1)
  expect(wrapper.vm.errors).toMatchObject([{ data: errorFixture }])
})
