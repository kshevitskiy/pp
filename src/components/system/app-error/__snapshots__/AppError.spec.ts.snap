// Vitest Snapshot v1

exports[`AppError component test suite > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`mount component 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
